src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -0,0 +1,182 @@
+            do
+            {
+                Stream stream = null;
+                IpcAdvertise advertise = null;
+                try
+                {
+                    stream = await _transport.AcceptAsync(linkedSource.Token);
+                }
+                catch (Exception ex) when (!(ex is OperationCanceledException))
+                {
+                    // The advertise data could be incomplete if the runtime shuts down before completely writing
+                    // the information. Catch the exception and continue waiting for a new connection.
+                }
+
+                // Cancel parsing of advertise data after timeout period to
+                // mitigate runtimes that write partial data and do not close the stream (avoid waiting forever).
+                using var parseCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(linkedSource.Token);
+                try
+                {
+                    parseCancellationSource.CancelAfter(TimeSpan.FromSeconds(3));
+
+                    advertise = await IpcAdvertise.ParseAsync(stream, parseCancellationSource.Token);
+                }
+                catch (OperationCanceledException ex) when (ex.CancellationToken == linkedSource.Token)
+                {
+                    // Only handle cancellation if it was due to the parse timeout.
+                }
+                catch (Exception ex) when (!(ex is OperationCanceledException))
+                {
+                    // The advertise data could be incomplete if the runtime shuts down before completely writing
+                    // the information. Catch the exception and continue waiting for a new connection.
+                }
+
+                if (null != advertise)
+                {
+                    Guid runtimeCookie = advertise.RuntimeInstanceCookie;
+                    int pid = unchecked((int)advertise.ProcessId);
+
+                    // If this runtime instance already exists, update the existing connection with the new endpoint.
+                    // Consumers should hold onto the connection instance and use it for diagnostic communication,
+                    // regardless of the number of times the same runtime instance connects. This requires consumers
+                    // to continuously invoke the AcceptAsync method in order to handle runtime instance reconnects,
+                    // even if the consumer only wants to handle a single connection.
+                    ServerIpcEndpoint endpoint = null;
+                    if (!_endpoints.TryGetValue(runtimeCookie, out endpoint))
+                    {
+                        // Create a new endpoint and connection that are cached an returned from this method.
+                        endpoint = new ServerIpcEndpoint();
+                        newConnection = new ReversedDiagnosticsConnection(this, endpoint, pid, runtimeCookie);
+
+                        _endpoints.TryAdd(runtimeCookie, endpoint);
```

Comment:
If this TryAdd fails, you may want to dispose the ServerIpcEndpoint and ReverseDiagnosticConnection that was just instantiated but failed to add.